include: autotools: do not symlink files in autoreconf

neocturne opened the pull request:
In Gluon's Github Actions CI, we were occasionally seeing bizarre build errors that looked like a config.sub file had been corrupted, or changed while it was being executed.

The cause turned out to be an interaction of the symlinks created by autoreconf (pointing from individual tools' build dirs into `staging_dir/host/share/automake-1.16`) and OpenWrt's host-build.mk, which replaced config.guess and config.sub *after* autoreconf. The result was that the replacement of these files ended up following the symlinks and writing the files in `staging_dir/host/share/automake-1.16` instead of a package's build dir. This could cause other packages' builds to fail if they were currently executing the scripts while they were being written.

To fix this, disable autoreconf's symlinking feature, so that modifying these files in a package's build directory can't accidentally affect the staged versions.

Signed-off-by: Matthias Schiffer <mschiffer@universe-factory.net>

neocturne committed:
include: autotools: do not symlink files in autoreconf

In Gluon's Github Actions CI, we were occasionally seeing bizarre build
errors that looked like a config.sub file had been corrupted, or changed
while it was being executed.

The cause turned out to be an interaction of the symlinks created by
autoreconf (pointing from individual tools' build dirs into
`staging_dir/host/share/automake-1.16`) and OpenWrt's host-build.mk,
which replaced config.guess and config.sub *after* autoreconf. The
result was that the replacement of these files ended up following the
symlinks and writing the files in `staging_dir/host/share/automake-1.16`
instead of a package's build dir. This could cause other packages' builds
to fail if they were currently executing the scripts while they were
being written.

To fix this, disable autoreconf's symlinking feature, so that modifying
these files in a package's build directory can't accidentally affect the
staged versions.

Signed-off-by: Matthias Schiffer <mschiffer@universe-factory.net>

neheb commented:
ping @mpratt14

mpratt14 commented:
I believe changing the way that scripts are copied would be a more appropriate fix. I have an old experimental branch where I worked on this part of the build system among other things. Let me know if you would like me to post a PR for a method similar to this, otherwise you're welcome to take this and adjust it as needed with my tag in the commit.

https://github.com/mpratt14/openwrt/commit/326bbc878d9834d9e4ee706c01c1490fba8cc9dc

for certain versions of cp, the `--remove-destinations` option should already accomplish solving the problem you're describing, but it seems to not be very portable... so even though we require coreutils to build, it may be best to rewrite the usage of xargs in order to both rm and cp each path.

https://unix.stackexchange.com/questions/667204/cp-command-follows-symbolic-link-at-destination-and-clobbers-link-target-instead

mpratt14 commented:
one thing that should be fixed in my commit which I just noticed is to use `?=` instead of just `=`...

neocturne commented:
I also thought about using `--remove-destination` and worried about portability to MacOS etc. However there are not just the common includes replacing `config.{guess,sub}`, but also individual Makefiles that do this (several of them in `toolchain/`.

Symlinking might be a good idea when the source files are installed at `/usr/share`, where a regular user can't modify them, but with OpenWrt staging dir, using symlinks seems risky and might introduce similar issues again in the future, even if we fix them now. I don't really see any downsides to just copying the files - the space these files take up should be negligible compared to the rest of the build dir.

mpratt14 commented:
In the individual Makefile cases for the toolchain it's not copying files into a symlink, they are just normal files before being overwritten because none of the toolchain specifically uses autoreconf. Anyway, I am ok with both types of solutions, but the main argument for yours should be to further properly isolate the build and staging directories, while fixing this race condition error that only happens sometimes is just one of perhaps several possible consequences.

It should be noted that neither of our solutions is guaranteed to keep this problem away forever, rather everyone should be vigilant to look out for a case where it can occur again. For example, we need to handle another case, for scripts like autogen.sh and bootstrap that call autoreconf in their own way that is not directly controlled in autotools.mk or elsewhere in the build system, but by the way those scripts are called, for example, when building libtool.

make sure to check this is an empty result after a full build:

`find build_dir -type l | xargs -n 1 ls -la | grep staging_dir`

mpratt14 commented:
so what method do we want to continue with? either one would need some small adjustment.

jow- commented:
> we need to handle another case, for scripts like autogen.sh and bootstrap that call autoreconf in their own way

At some point in the past it was deemed to be a packaging bug if something invoked autogen.sh or bootstrap kinds of scripts during build. See for example commits https://github.com/openwrt/packages/commit/dcb2407c89638b98eca547fa766f38c2179fbd8d or https://github.com/openwrt/packages/commit/2e92e83d10a576b3be7ba6b228b45bdaefe9cb18 or https://github.com/openwrt/packages/commit/f957dfcb04bc3e3f1697c1507a9962506931ed68

mpratt14 commented:
I agree that it's not ideal at all to have autogen.sh used in builds, so perhaps someone can look into finding a way to remove that from genext2fs.

However, when available, the bootstrap script is the best way to incorporate gnulib sources into the build (compared to gnulib-tool), so for those cases it just needs a flag added to it so it copies instead of symlinks.

neocturne commented:
Also, for packages from Git sources, running an `autogen.sh` or similar script is often the intended/only way to generate certain build scripts - such a script may do more than just call autoreconf. I would *expect* that these scripts do not create any symlinks pointing outside the project by default, because they are usually supposed to create something that is distributable as a tarball. Using OpenWrt's autoreconf support should still be preferred when it's all that's needed.

In any case, I'll run another build and verify that there aren't any other symlinks pointing into the staging dir as suggested by @mpratt14.

neocturne committed:
tools: libtool: do not symlink files in bootstrap

Another instance of files in build_dir symlinking to staging_dir. While
the symlinks do not currently cause any bugs in the libtool package,
such symlinks were found to make the build more fragile, as writing to
the symlink may accidentally modify the shared file in staging_dir. Pass
--copy to bootstrap to disable the symlinking.

Signed-off-by: Matthias Schiffer <mschiffer@universe-factory.net>

neocturne commented:
Okay, I have pushed another commit that also disables the symlinking for libtool's bootstrap. I haven't done a "full" build of all packages, but only a pretty much default build for a single target; with this, only a small number of symlinks from `build_dir` to other parts of the OpenWrt directory remain:

- `build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/u-boot-fritz4040/u-boot-2022.05.01~9d89013f/fritz/lzma -> .../staging_dir/host/bin/lzma` - this is the only symlink into `staging_dir` now. It seems harmless, as the symlink is only ever created at one line of the U-Boot Makefiles, and never written by anything else
- A number of packages have a symlink to `.../scripts/config.rpath`. Again, this is a file specific to OpenWrt's build system, to chances for other code to mangle it are low

Do you have an opinion regarding these two files? Is it okay to leave them as symlinks, or should they be copied as well?

neocturne commented:
@mpratt14 Do you have further comments regarding this PR? If not, I'll merge this soon, as it fixes a build issue we're seeing in Gluon's CI (rarely, but frequently enough to be annoying)

mpratt14 commented:
it's good enough for now I suppose 

maybe I can handle the case of genext2fs myself if no one else does as long as I remember to do it

neocturne commented:
Merged to main.

ynezz commented:
@neocturne FYI it seems to be causing some issues with at least `ntpd` package https://github.com/openwrt/packages/issues/24918

neocturne commented:
@ynezz Interesting! This might mean that the package previously only built because some of the files used by autoreconf were overwritten by accident. I'll try to have a look some time this week.
